### PR TITLE
Switch to using gradle artifact APIs for root project tasks

### LINKED ISSUE #701: Switch Skippy avoided projects recording to configurations
In order to support Project Isolation, we cannot have subprojects reach directly into `rootProject.tasks`. Instead, we need to have these projects produce consumable configurations that the root project consumes.

- Docs: https://docs.gradle.org/current/userguide/cross_project_publications.html
- Example (Jacoco): https://docs.gradle.org/current/userguide/jacoco_plugin.html#sec:outgoing_variants

**API Shape**

Producing subproject
```kotlin
plugins {
  // ...
  id("slack.gradle.skippy")
}
```

Consuming root project
```kotlin
plugins {
  // ...
  id("slack.gradle.skippy.root")
}

dependencies {
  for (subproject in subprojects) {
    skippyProject(subproject)
  }
}
```

### PR DESCRIPTION
**Pull request by ZacSweers**
Gradle's upcoming Project Isolation feature prohibits direct access of other projects' tasks (including the root project). This means a lot of our plumbing for things like skippy, mod stats, and androidTest APKs are incompatible, as they all use a pattern of reaching into the rootProject to get aggregating tasks.

This PR updates all those implementations to use Gradle's artifacts APIs to publish artifacts from these subproject tasks and consume them via dependency configurations in the root project.

Resolves #701, and multiple others like it that I found along the way.

More reading: https://docs.gradle.org/current/userguide/cross_project_publications.html#sec:variant-aware-sharing

### CODE REVIEW

File: slack-plugin/src/main/kotlin/slack/dependencyrake/DependencyRake.kt [hunk 1]
```diff
@@ -459,7 +461,10 @@ internal abstract class MissingIdentifiersAggregatorTask : DefaultTask() {
     const val NAME = "aggregateMissingIdentifiers"
 
     fun register(rootProject: Project): TaskProvider<MissingIdentifiersAggregatorTask> {
+      val resolver =
+        Resolver.interProjectResolver(rootProject, SgpArtifacts.Kind.DAGP_MISSING_IDENTIFIERS)
       return rootProject.tasks.register(NAME, MissingIdentifiersAggregatorTask::class.java) {
+        inputFiles.from(resolver.artifactView())
```
Comment (autonomousapps) [inline slack-plugin/src/main/kotlin/slack/dependencyrake/DependencyRake.kt:467]: oh I like this, I will probably steal it

File: slack-plugin/src/main/kotlin/slack/gradle/StandardProjectConfigurations.kt [hunk 2]
```diff
@@ -505,12 +500,22 @@ internal class StandardProjectConfigurations(
...
+              // the same output
+              SimpleFileProducerTask.registerOrConfigure(
+                  project,
+                  name = "androidTestProjectMetadata",
+                  description =
+                    "Produces a metadata artifact indicating this project path produces an androidTest APK.",
+                  input = projectPath,
+                  group = "skippy"
+                )
+                .publishWith(skippyAndroidTestProjectPublisher)
```
Comment (autonomousapps) [inline slack-plugin/src/main/kotlin/slack/gradle/StandardProjectConfigurations.kt:513]: nice

File: slack-plugin/src/main/kotlin/slack/gradle/GradleExt.kt [hunk 3]
```diff
@@ -321,3 +321,13 @@ internal inline fun <reified T : Any> Project.serviceOf(): T =
...
+
+@Suppress("UNCHECKED_CAST")
+internal inline fun <reified T : Task> TaskContainer.registerOrConfigure(
+  taskName: String,
+  crossinline configureAction: T.() -> Unit
+): TaskProvider<T> =
+  when (taskName) {
+    in names -> named(taskName) as TaskProvider<T>
+    else -> register(taskName, T::class.java)
+  }.apply { configure { configureAction() } }
```
Comment (autonomousapps) [inline slack-plugin/src/main/kotlin/slack/gradle/GradleExt.kt:333]: this is a neat pattern

File: slack-plugin/src/main/kotlin/slack/gradle/StandardProjectConfigurations.kt [hunk 4]
```diff
@@ -588,12 +593,12 @@ internal class StandardProjectConfigurations(
...
-                // time
-                // dependencies start getting resolved.
+                // time dependencies start getting resolved.
                 //
-                slackTools.globalConfig.updateRobolectricJarsTask?.let {
+                slackProperties.versions.robolectric?.let {
                   logger.debug("Configuring $name test task to depend on Robolectric jar downloads")
-                  test.dependsOn(it)
+                  // Depending on the root project task by name alone is ok for Project Isolation
+                  test.dependsOn(UpdateRobolectricJarsTask.NAME)
```
Comment (autonomousapps) [inline slack-plugin/src/main/kotlin/slack/gradle/StandardProjectConfigurations.kt:601]: oh gradle 👨‍🍳 💋

File: slack-plugin/src/main/kotlin/slack/gradle/artifacts/Publisher.kt [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+ * Gradle uses [attributes][Attr] to wire the consumer project's [Resolver.internal] (resolvable)
+ * configuration to the producer project's [external] (consumable) configuration, which is itself
+ * configured via [publish].
+ *
+ * @see <a
+ *   href="https://docs.gradle.org/current/userguide/cross_project_publications.html#sec:variant-aware-sharing">Variant-aware
+ *   sharing of artifacts between projects</a>
+ * @see <a
+ *   href="https://dev.to/autonomousapps/configuration-roles-and-the-blogging-industrial-complex-21mn">Gradle
+ *   configuration roles</a>
```
Comment (autonomousapps) [inline slack-plugin/src/main/kotlin/slack/gradle/artifacts/Publisher.kt:49]: oh that's kind of nice. it always annoyed me that having these on a single line was so long, but this looks readable even in github

File: slack-plugin/src/main/kotlin/slack/gradle/artifacts/Publisher.kt [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    external.configure { outgoing.artifact(output) }
+  }
+
+  /**
+   * Teach Gradle which thing produces the artifact associated with the external/consumable
+   * configuration.
+   */
+  fun publishDirs(output: Provider<Directory>) {
+    external.configure { outgoing.artifact(output) }
+  }
```
Comment (autonomousapps) [inline slack-plugin/src/main/kotlin/slack/gradle/artifacts/Publisher.kt:110]: you could always pull a Gradle and do `fun publish(output: Any)` :trollface:

File: slack-plugin/src/main/kotlin/slack/gradle/artifacts/Resolver.kt [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+  fun addSubprojectDependencies(slackProperties: SlackProperties) {
+    project.dependencies.apply {
+      for (subproject in project.subprojects) {
+        if (subproject.path == slackProperties.platformProjectPath) continue
+        add(declarable.name, project.project(subproject.path))
+      }
+    }
+  }
```
Comment (autonomousapps) [inline slack-plugin/src/main/kotlin/slack/gradle/artifacts/Resolver.kt:107]: _takes notes_

File: slack-plugin/src/main/kotlin/slack/gradle/artifacts/SgpArtifacts.kt [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+  ) {
+    SKIPPY_UNIT_TESTS("skippyUnitTests", "skippy-unit-tests"),
+    SKIPPY_LINT("skippyLint", "skippy-lint"),
+    SKIPPY_AVOIDED_TASKS("skippyAvoidedTasks", "skippy-avoided-tasks"),
+    SKIPPY_ANDROID_TEST_PROJECT("skippyAndroidTestProject", "skippy-android-test-project"),
+    SKIPPY_DETEKT("skippyDetekt", "skippy-detekt"),
+    ANDROID_TEST_APK_DIRS("androidTestApkDirs", "android-test-apk-dirs"),
+    DAGP_MISSING_IDENTIFIERS("dagpMissingIdentifiers", "dagp-missing-identifiers"),
+    MOD_STATS_STATS_FILES("modStatsFiles", "mod-stats-files"),
+  }
```
Comment (autonomousapps) [inline slack-plugin/src/main/kotlin/slack/gradle/artifacts/SgpArtifacts.kt:40]: if I were to try to turn this into a library, what are your thoughts on how I'd handle this sort of thing? My project and your project have different artifact kinds, but the `SgpArtifacts` type is baked into the resolver/publisher APIs.

File: slack-plugin/src/main/kotlin/slack/gradle/tasks/FileTasks.kt
Comment (autonomousapps) [inline slack-plugin/src/main/kotlin/slack/gradle/tasks/FileTasks.kt:1]: This whole pattern seems really nice and ergonomic.

**Review by autonomousapps — APPROVED** [review]
LGTM.

Would you mind if I provided an updated license header though?

**Comment by ZacSweers** [comment]
> Would you mind if I provided an updated license header though?

Yep will add

File: slack-plugin/src/main/kotlin/slack/gradle/artifacts/SgpArtifacts.kt (hunk 8 above)
Comment (ZacSweers) [reply slack-plugin/src/main/kotlin/slack/gradle/artifacts/SgpArtifacts.kt]: I updated the PR with a sketch of a more agnostic API, but it forgoes use of `Named` and instead expects the artifact you give it to implement equals(). This takes a page out of AGP's artifact handling (i.e. like `SingleArtifact`'s subtypes). I think this works pretty well

File: slack-plugin/src/main/kotlin/slack/gradle/artifacts/SgpArtifact.kt [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+  final override val attribute: Attribute<SgpArtifact>
+    get() = SGP_ARTIFACTS_ATTRIBUTE
+
+  companion object {
+    @JvmField
+    val SGP_ARTIFACTS_ATTRIBUTE: Attribute<SgpArtifact> =
+      Attribute.of("sgp.internal.artifacts", SgpArtifact::class.java)
+  }
+
+  data object SKIPPY_UNIT_TESTS : SgpArtifact("skippyUnitTests") {
```
Comment (autonomousapps) [inline slack-plugin/src/main/kotlin/slack/gradle/artifacts/SgpArtifact.kt:32]: TIL `data object`

**Pull request merged**